sinker/config_state_check.go

```diff
@@ -0,0 +1,56 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+package sinker
+
+import (
+	"github.com/ns1labs/orb/sinker/config"
+	"go.uber.org/zap"
+	"time"
+)
+
+const (
+	streamID       = "orb.sinker"
+	streamLen      = 1000
+	CheckerFreq    = 300 * time.Second
+	DefaultTimeout = 30 * time.Minute
+)
+
+func (svc *sinkerService) checkState(t time.Time) {
+	owners, err := svc.sinkerCache.GetAllOwners()
+	if err != nil {
+		svc.logger.Error("failed to retrieve the list of owners")
+		return
+	}
+
+	for _, ownerID := range owners {
+		configs, err := svc.sinkerCache.GetAll(ownerID)
+		if err != nil {
+			svc.logger.Error("unable to retrieve policy state", zap.Error(err))
+			return
+		}
+		for _, cfg := range configs {
+			// Set idle if the sinker is more then 30 minutes not been sending metrics
+			if cfg.LastRemoteWrite.Add(DefaultTimeout).Before(time.Now()) {
```

Comment:
So this doesn't work (well) because every instance of sinker in the cluster will be duplicating the timeout calculations and updates. The reason it works better for Agents is because the "last timestamp" state and the action for setting "stale" status is done in postgres - all instances in the cluster will just call the function to update the status in the database, and only the first one will actually do anything. In this code, all instances will all write to redis.  

I think the better way to do this is to do what we do with Agents, and add a new 'ts_last_sink' field to postgres which is updated upon successful sink. Then this function can do what agent does, and call the repo function to update all 'active' sinks with a 'ts_last_sink' older than DefaultTimeout to 'idle'

Reply:
ts_last_sink should be a field from what database? I'm using lastRemoteWrite, which is a timestamp field, stored on Redis to check the last sink time. I guess it's working the same way the ts_last_sink would work on Postgres. I've changed the function to only set idle for active configurations, so the first one is going the change the value to idle if we have more than one instance. I'm asking all this because we don't have a Postgres instance for sinker service and I'm using Redis to store the configs...

Reply:
Yeah it would be the sinks database/table so this "state check" functionality would have to move there instead of sinker. I do see there would be some complication in keeping that up to date when sinker writes metrics - it would have to go through the event stream.

Ok if we leave this code in place - I'm still concerned that all instances will all iterate through all sinks in redis every couple of minutes. How about if instead of changing the state to 'idle' in redis, we removed it, so that we _only_ store "active" sinks in redis? Presumably when that sink starts receiving metrics again, it updates itself to "active" and recaches?